core/src/main/scala/org/apache/spark/util/EventLoop.scala

```
@@ -0,0 +1,110 @@
+
+    override def run(): Unit = {
+      try {
+        while (true) {
+          val event = eventQueue.take()
+          try {
+            onReceive(event)
+          } catch {
+            case NonFatal(e) => {
+              try {
+                onError(e)
+              } catch {
+                case NonFatal(e) => logError("Unexpected error in " + name, e)
+              }
+            }
+          }
+        }
+      } catch {
+        case ie: InterruptedException => // exit even if eventQueue is not empty
+        case NonFatal(e) => logError("Unexpected error in " + name, e)
+      }
+    }
+
+  }
+
+  def start(): Unit = {
+    // Call onStart before starting the event thread to make sure it happens before onReceive
+    onStart()
+    eventThread.start()
+  }
+
+  def stop(): Unit = {
+    eventThread.interrupt()
+    eventThread.join()
+    // Call onStop after the event thread exits to make sure onReceive happens before onStop
+    onStop()
+  }
+
+  /**
+   * Put the event into the event queue. The event thread will process it later.
+   */
+  def post(event: E): Unit = {
+    eventQueue.put(event)
+  }
+
+  /**
+   * Return if the event thread has already been started but not yet stopped.
+   */
+  def isActive: Boolean = eventThread.isAlive
+
+  /**
+   * Invoke when `start()` is called. It's also invoked before the event thread starts.
+   */
+  def onStart(): Unit = {}
+
+  /**
+   * Invoke when `stop()` is called and the event thread exits.
+   */
+  def onStop(): Unit = {}
+
+  /**
+   * Invoke in the event thread when polling events from the event queue.
+   *
+   * Note: Should avoid calling blocking actions in `onReceive`, or the event thread will be blocked
+   * and cannot process events in time. If you want to call some blocking actions, run them in
+   * another thread.
+   */
+  def onReceive(event: E): Unit
+
+  /**
+   * Invoke if `onReceive` throws any non fatal error. `onError` must not throw any non fatal error.
```

The second condition is not really true, right? Such errors are simply ignored, so I don't think we need to make a stipulation against them.